### PR TITLE
[v0.10.3] service analyser improvements

### LINKED ISSUE #1898: Service analyzer improvements
### Subject of the issue
The service analysis command is currently difficult to use. At the moment, I have detected the following issues:

#### Confusing error message
Running the service analyzer:

```
autonomy analyse service --public-id valory/oracle_hardhat
```

with [consensus_threshold specified](https://github.com/valory-xyz/price-oracle/blob/fe74448d8c8dfc28936212641868dd902be64734/packages/valory/skills/oracle_abci/skill.yaml#L138) on the skill configuration, gives:

```
[2023-03-24 09:49:06,371][INFO] Validating service overrides
[2023-03-24 09:49:06,371][INFO] Validating (skill, valory/oracle_abci:0.1.0) from the service overrides
Error: ABCI skill validation failed; 'consensus_threshold' is a required property
```

That is because the *service* does not specify the `consensus_threshold`, but the message incorrectly reports that it is missing from the skill.

#### Configurations required in setup are reported as missing

If we go ahead and add the `consensus_threshold` in the service configuration, but do not place it under the `setup`, then it is reported as missing, instead of mentioning that the `consensus_threshold` was not found in the `setup` configuration.

#### Incorrectly reporting as missing

If we [fix all the issues](https://github.com/valory-xyz/price-oracle/pull/57/commits/e2519777a136213a9824bf8c6e0b96198b216945) reported by the tool, then it incorrectly performs the cross-verification:

```
[2023-03-24 11:03:51,366][INFO] Cross verifying overrides between skill and agent
[2023-03-24 11:03:51,366][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (skill, valory/oracle_abci:0.1.0)
        Path: models
        Missing parameters: {'price_api', 'randomness_api', 'server_api'}
```

#### Too verbose

The tool is too verbose. Even though everything is correct in the configurations on [price-oracle@e2519777a136213a9824bf8c6e0b96198b216945](https://github.com/valory-xyz/price-oracle/pull/57/commits/e2519777a136213a9824bf8c6e0b96198b216945), the tool reports all the following:

```
[2023-03-24 11:03:51,361][INFO] Validating service overrides
[2023-03-24 11:03:51,362][INFO] Validating (skill, valory/oracle_abci:0.1.0) from the service overrides
[2023-03-24 11:03:51,362][INFO] Validating (connection, valory/ledger:0.19.0) from the service overrides
[2023-03-24 11:03:51,362][INFO] Validating agent overrides
[2023-03-24 11:03:51,362][INFO] Validating skill overrides
[2023-03-24 11:03:51,362][INFO] Cross verifying overrides between agent and service
[2023-03-24 11:03:51,363][INFO] Verifying the skill override with index 0
[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args
        Missing parameters: {'response_index'}

[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args.headers
        Missing parameters: {'$', 'c', '}', 'd', 'i', '{', 't', ':'}

[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args.parameters
        Missing parameters: {'pair', 'symbol'}

[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args
        Missing parameters: {'tendermint_com_url', 'termination_sleep', 'reset_tendermint_after', 'multisend_address', 'validate_timeout', 'init_fallback_gas', 'tendermint_url', 'round_timeout_seconds', 'use_termination', 'max_healthcheck', 'tendermint_p2p_url'}

[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args.setup
        Missing parameters: {'oracle_contract_address'}

[2023-03-24 11:03:51,364][INFO] Verifying the skill override with index 1
[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args
        Missing parameters: {'headers'}

[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args.parameters
        Missing parameters: {'ids', 'vs_currencies', 'symbol'}

[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args
        Missing parameters: {'tendermint_com_url', 'termination_sleep', 'reset_tendermint_after', 'multisend_address', 'validate_timeout', 'init_fallback_gas', 'tendermint_url', 'round_timeout_seconds', 'use_termination', 'max_healthcheck', 'tendermint_p2p_url'}

[2023-03-24 11:03:51,364][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args.setup
        Missing parameters: {'oracle_contract_address'}

[2023-03-24 11:03:51,364][INFO] Verifying the skill override with index 2
[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args
        Missing parameters: {'response_index'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args.headers
        Missing parameters: {'$', 'c', '}', 'd', 'i', '{', 't', ':'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args.parameters
        Missing parameters: {'vs_currencies', 'ids', 'pair', 'symbol'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args
        Missing parameters: {'tendermint_com_url', 'termination_sleep', 'reset_tendermint_after', 'multisend_address', 'validate_timeout', 'init_fallback_gas', 'tendermint_url', 'round_timeout_seconds', 'use_termination', 'max_healthcheck', 'tendermint_p2p_url'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args.setup
        Missing parameters: {'oracle_contract_address'}

[2023-03-24 11:03:51,365][INFO] Verifying the skill override with index 3
[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args
        Missing parameters: {'response_index'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args.headers
        Missing parameters: {'$', 'c', '}', 'd', 'i', '{', 't', ':'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.price_api.args.parameters
        Missing parameters: {'ids', 'vs_currencies', 'pair'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args
        Missing parameters: {'tendermint_com_url', 'termination_sleep', 'reset_tendermint_after', 'multisend_address', 'validate_timeout', 'init_fallback_gas', 'tendermint_url', 'round_timeout_seconds', 'use_termination', 'max_healthcheck', 'tendermint_p2p_url'}

[2023-03-24 11:03:51,365][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (service, valory/oracle_hardhat:0.1.0)
        Path: models.params.args.setup
        Missing parameters: {'oracle_contract_address'}

[2023-03-24 11:03:51,366][INFO] Cross verifying overrides between skill and agent
[2023-03-24 11:03:51,366][WARNING] (agent, valory/oracle:0.1.0) contains configuration which is missing from (skill, valory/oracle_abci:0.1.0)
        Path: models
        Missing parameters: {'price_api', 'randomness_api', 'server_api'}

[2023-03-24 11:03:51,366][WARNING] (skill, valory/oracle_abci:0.1.0) contains configuration which is missing from (agent, valory/oracle:0.1.0)
        Path: models.params.args
        Missing parameters: {'cleanup_history_depth_current', 'oracle', 'genesis_config', 'retry_attempts', 'drand_public_key', 'request_retry_delay', 'cleanup_history_depth', 'tendermint_max_retries', 'observation_aggregator_function', 'tendermint_check_sleep_delay', 'finalize_timeout', 'tx_timeout', 'history_check_timeout', 'retry_timeout', 'keeper_timeout', 'request_timeout', 'keeper_allowed_retries', 'sleep_time', 'service_id', 'max_attempts'}

Service is ready to be deployed.
```

This makes it extremely hard to understand if something is going wrong.

#### Not being able to run the command on CI defeats its purpose

We should add the tool to our workflows, otherwise, we can still end up deploying a problematic service by human error.

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

This PR

- Implements custom schema validator to report all validation issues at once
- Adds support for skip warnings
- Raises warning when components are defined in the agent config and not in the service config
- improves error messages
- Adds support for validating environment overrides

## Fixes

#1898 
#1873

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have locally run services that could be impacted and they do not present failures derived from my changes
- [ ] Public-facing documentation has been updated with the changes affected by this PR. Even if the provided contents are not in their final form, all significant information must be included.
- [ ] Any backwards-incompatible/breaking change has been clearly documented in the upgrading document.

### CODE REVIEW

File: autonomy/analyse/service.py [hunk 1]
```diff
@@ -319,16 +376,38 @@ def cross_verify_overrides(
...
-                f"packages with missing overrides={missing_from_agent}"
+                "\n\t- ".join(
+                    map(
+                        str,
+                        [
+                            "Overrides with following component IDs are defined in the service configuration but not in the agent configuration",
+                            *missing_from_agent,
+                        ],
+                    )
+                )
```
Comment (angrybayblade) [inline autonomy/analyse/service.py:404]: Addresses https://github.com/valory-xyz/open-autonomy/issues/1898#issuecomment-1482788781

File: autonomy/analyse/service.py [hunk 2]
```diff
@@ -319,16 +376,38 @@ def cross_verify_overrides(
...
             )
 
-        skill_override_from_agent = agent_config.component_configurations[
-            skill_config.package_id
-        ]
+        skill_override_from_agent = apply_env_variables(
+            agent_config.component_configurations[skill_config.package_id],
+            env_variables=os.environ.copy(),
+        )
```
Comment (angrybayblade) [inline autonomy/analyse/service.py:410]: This was a bug

File: autonomy/analyse/service.py [hunk 3]
```diff
@@ -375,6 +455,117 @@ def cross_verify_overrides(
...
         )
 
+    @classmethod
+    def validate_override_env_vars(
+        cls,
+        overrides: Union[OrderedDict, dict],
+        validate_env_var_name: bool = False,
+        json_path: Optional[List[str]] = None,
+    ) -> List[str]:
+        """Validate environment variables."""
```
Comment (angrybayblade) [inline autonomy/analyse/service.py:465]: This check will make sure that

- All overrides are defined as the env vars at agent layer
- All overrides are defined as the env vars with env var name defined at the service layer
- Env vars can be loaded, data types and default values are defined properly

and prevent issues like this https://github.com/valory-xyz/open-autonomy/issues/1898#issuecomment-1493139762 and https://github.com/valory-xyz/open-autonomy/issues/1898#issuecomment-1497256903

File: autonomy/analyse/service.py [hunk 4]
```diff
@@ -383,104 +574,160 @@ def _validate_override(
...
+            self._validate_override(
                 validator=ABCI_SKILL_VALIDATOR,
                 overrides=override,
                 has_multiple_overrides=has_multiple_overrides,
-                error_message="ABCI skill validation failed; {error}",
+                error_message=(
+                    "ABCI Skill `"
+                    + str(component_id.public_id)
+                    + "` override validation failed; {error}"
+                ),
```
Comment (angrybayblade) [inline autonomy/analyse/service.py:621]: Fixes https://github.com/valory-xyz/open-autonomy/issues/1898#issuecomment-1493118496

File: autonomy/analyse/service.py [hunk 5]
```diff
@@ -168,19 +171,62 @@
...
+
+    def validate(self, *args: Any, **kwargs: Any) -> None:
+        """Validate and raise all errors at once."""
+        missing = []
+        extra = []
+        for error in self.iter_errors(*args, **kwargs):
+            message = cast(ValidationError, error).message
+            missing_property_check = REQUIRED_PROPERTY_RE.match(message)
+            if missing_property_check is not None:
+                (missing_property,) = cast(re.Match, missing_property_check).groups()
```
Comment (solarw) [inline autonomy/analyse/service.py:199]: this unpack works? or `*_` is required?

File: autonomy/analyse/service.py [hunk 6]
```diff
@@ -168,19 +171,62 @@
...
+
+
+class CustomSchemaValidator(Draft4Validator):
+    """Custom schema validator to report all missing fields at once."""
+
+    def validate(self, *args: Any, **kwargs: Any) -> None:
+        """Validate and raise all errors at once."""
+        missing = []
+        extra = []
+        for error in self.iter_errors(*args, **kwargs):
```
Comment (solarw) [inline autonomy/analyse/service.py:195]: why not create ValidationError with list of all errors strings and show to user as is, without grouping? will make code more simple

File: autonomy/analyse/service.py [hunk 7]
```diff
@@ -168,19 +171,62 @@
...
+
+            property_not_allowed_check = PROPERTY_NOT_ALLOWED_RE.match(message)
+            if property_not_allowed_check is not None:
+                extra += re.findall(
+                    "[a-zA-Z0-9]+", property_not_allowed_check.groups()[0]
+                )
+                continue
+
+            if "does not have enough properties" in message:
+                raise error
```
Comment (solarw) [inline autonomy/analyse/service.py:211]: should it be stored and showed with other errors?

File: autonomy/analyse/service.py [hunk 8]
```diff
@@ -168,19 +171,62 @@
...
+
+        if len(extra) > 0:
+            raise ValidationError(
+                message="Additional properties found "
+                + ", ".join(map(lambda x: f"'{x}'", extra))
+            )
+
+        if len(missing) > 0:
+            raise ValidationError(
```
Comment (solarw) [inline autonomy/analyse/service.py:220]: two dedicated raises here (actually 3 in the function body)
looks like does not meet requirement show all the errors at once, but show all the same group errors at once

**Review by solarw — COMMENTED** [review]
some points on validation errors grouping

File: autonomy/analyse/service.py (hunk 5 above)
Comment (angrybayblade) [reply autonomy/analyse/service.py:199]: There's only one group so this should work

File: autonomy/analyse/service.py (hunk 6 above)
Comment (angrybayblade) [reply autonomy/analyse/service.py:195]: To do this we will have to create multiple error string for each missing and each additional property which will defeat the purpose of having this class for reporting all missing and additional properties at once

Comment (solarw) [reply autonomy/analyse/service.py:238]: @angrybayblade 
we can create custom exception with all errors incapsulated and on cli side print this error message properly line by line

Comment (angrybayblade) [reply autonomy/analyse/service.py:238]: I tried to address it here [1d63285](https://github.com/valory-xyz/open-autonomy/pull/1942/commits/1d63285e9626f4df1dad827b66896848a36b420f)

**Comment by Adamantios** [comment]
Looks good, let's add `tox -e analyse-services` to the workflow.

**Review by solarw — DISMISSED** [review]
LGTM.
nice work!

**Pull request merged**